feat(localize): support merging multiple translation files

petebacondarwin opened the pull request:
Previously only one translation file per locale could be loaded.
Now the user can specify multiple files per locale, and the translations
from each of these files will be merged together by message id.

---

Note that the CLI uses the lower level `TranslationParser` classes directly, so will not inherit this new behaviour automatically - cc @clydin.

petebacondarwin committed:
feat(localize): support merging multiple translation files

Previously only one translation file per locale could be loaded.

Now the user can specify multiple files per locale, and the translations
from each of these files will be merged together by message id.
The merging is on a first-wins approach. So if to you have three files
to be merged:

```
['a.xlf', 'b.xmb', 'c.json']
```

Then any message from `a.xlf` will be used rather than a message from `b.xmb`
or `c.json` and so on. In practice this means that you should put the files
in order of most important first, with "fallback" translations later.

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_loader.ts:
```diff
@@ -31,46 +31,87 @@ export class TranslationLoader {
-    });
+      Object.keys(nextBundle.translations).forEach(messageId => {
+        if (bundle.translations[messageId] !== undefined) {
+          this.diagnostics?.error(
```
@petebacondarwin I'm wondering if this behavior should be configurable (a-la https://angular.io/guide/i18n#report-missing-translations), but default to "error". An example I can think of is if you want to use a 3rd party component in your app and this component also includes translations, but you want to change some of the messages. In this case it might be convenient to provide component's translation file and your own translation file that overrides some of the messages.

What do you think about it?

petebacondarwin replied:
Actually that sounds like a very common scenario... Perhaps we should default to "warn"?

petebacondarwin replied:
I added options to change whether this errors or not and defaulted it to warning.
(also took the opportunity to add a refactor commit to the start of the PR)

AndrewKushnir reviewed (approved):
LGTM, thanks @petebacondarwin 👍 Just left one question/proposal to consider.

petebacondarwin committed:
refactor(localize): simplify adding condition diagnostics

Previously the `missingTranslation` option had to be checked before calling
`warn` or `error` on the `diagnostics` object.

Now this boilerplate is hidden inside the `Diagnostics.add()` method, which
will open it up to being used for other conditional diagnostics.

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_loader.ts:
```diff
@@ -16,6 +17,7 @@ import {TranslationParser} from './translation_parsers/translation_parser';
 export class TranslationLoader {
   constructor(
       private translationParsers: TranslationParser<any>[],
+      private duplicateTranslation: DiagnosticHandlingStrategy = 'warning',
```
nit: I think `TranslationLoader` should not make a decision on how to handle duplicate translations if nothing is provided. I'd propose to set default value in a different place, where we call `new TranslationLoader`.

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_loader.ts:
```diff
@@ -31,46 +33,88 @@ export class TranslationLoader {
+
+      // If we were passed a diagnostics object then copy the messages over to it.
+      if (this.diagnostics) {
+        this.diagnostics.messages.push(...diagnostics.messages);
```
The `messages` is marked as `readonly` on the `Diagnostics` class (which doesn't prevent new items from being added) so I'm wondering if adding a method to `Diagnostics` class would be better (and also making `messages` a private field)?

petebacondarwin replied:
Yeah, this is how it was before this PR too. I'll see what I can do.

petebacondarwin replied:
I added a `merge()` method but I would like to keep the `messages` property public since this is how one reads the messages out of the `Diagnostics` object. Adding a getter property or method to hide this does not provide much better encapsulation, unless we actually clone the `messages` array when accessing it, which seems over the top. WDYT?

AndrewKushnir commented on packages/localize/src/tools/src/translate/main.ts:
```diff
@@ -117,7 +129,7 @@ export interface TranslateFilesOptions {
    * working directory.
    */
-  translationFilePaths: string[];
+  translationFilePaths: (string|string[])[];
```
I think it'd be great to add a note regarding duplicate translations rule, i.e. if we have 2 bundles that contain translations with the same id, what'd happen in this case (translation from the first bundle would be taken or translation from the second bundle will override the first one).

petebacondarwin replied:
Will do!

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_loader.ts:
```diff
@@ -31,46 +33,88 @@ export class TranslationLoader {
-    });
+      Object.keys(nextBundle.translations).forEach(messageId => {
+        if (bundle.translations[messageId] !== undefined) {
+          this.diagnostics?.add(
```
Based on this logic, if we have multiple bundles and same translation is present in multiple bundles, the first translation wins. Intuitively, this problem is similar to merging objects, for example `{...a, ...b}`, so `b` overrides keys also present in `a`, i.e. the last object has higher priority. Pete, WDYT?

petebacondarwin replied:
First translation wins is the approach I am aiming for. The conceptual model being that you provide the application translation file first and then "fallback" library translation files after:

```
[ 'app.en.xlf', 'lib1.en.xlf', 'lib2.en.xlf]
```

I will improve the commit message and documentation comments with this.

AndrewKushnir reviewed (commented):
Thanks for the updates Pete! I've left a few comments, could you please have a look when you get a chance?

petebacondarwin committed:
refactor(localize): hide merging diagnostics messages behind a method

This commit introduces `Diagnostics.merge(other)` which will take the messages from
`other` and append them to the messages of `this`.

The translation loader is updated to use this new method.

mhevery reviewed (approved):
Integration tests LGTM

AndrewKushnir reviewed (approved):
LGTM, thanks Pete (also special thanks for the docs!) 👍
